### PR TITLE
Added Lucas theorem for Java and Python

### LINKED ISSUE #1081: Lucas Theorem in Python and Java
I see Lucas Theorem is available in c and c++. Can I start working on Python and Java?
Link to the folder: [Lucas Algo](https://github.com/jainaman224/Algo_Ds_Notes/tree/master/Lucas%20Theorem)

### PR DESCRIPTION
**Pull request by prashant-mahanta**
Fixes #1081

#### Checklist:
- [x] 4 space indentation.
- [x] Coding conventions are followed.
- [x] Input is taken dynamically.
- [x] Sample Input / Output is added at the end of file.
- [x] Logic Documentation (Comments).
- [x] File names are correct.

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

#### Changes proposed in this pull request:
-  Lucas Theorem in Java and Python

#### Languages Used:
-  Java
- Python

#### Files Added:
- Lucas_Theorem.java
- Lucas_Theorem.py

> We're happy to help you get this ready -- don't be afraid to ask for help.

Thanks!

### CODE REVIEW

File: Lucas_Theorem/Lucas_Theorem.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
+import java.util.*;
+public class Lucas_Theorem {
```
Comment (MastersAbh) [inline Lucas_Theorem/Lucas_Theorem.java]: A space before class declaration would be nice.

File: Lucas_Theorem/Lucas_Theorem.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+    static int nCrModpDP (int n, int r, int p)
+    {
+        int C[] = new int[r+1];
```
Comment (MastersAbh) [inline Lucas_Theorem/Lucas_Theorem.java]: A space around +

File: Lucas_Theorem/Lucas_Theorem.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        C[0] = 1;
+
+        for (int i=1; i<=n; i++)
```
Comment (MastersAbh) [inline Lucas_Theorem/Lucas_Theorem.java]: please add space around operators.

File: Lucas_Theorem/Lucas_Theorem.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        {
+            for(int j=Math.min(i,r); j>0; j--)
```
Comment (MastersAbh) [inline Lucas_Theorem/Lucas_Theorem.java]: please add space around operators.

Comment (MastersAbh) [reply Lucas_Theorem/Lucas_Theorem.java]: also around comma

File: Lucas_Theorem/Lucas_Theorem.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+                C[j] = (C[j] + C[j-1]) % p;
```
Comment (MastersAbh) [inline Lucas_Theorem/Lucas_Theorem.java]: please add space around operators.

File: Lucas_Theorem/Lucas_Theorem.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+        }
+
+        return C[r];
+    }
+
+    static int nCrModpLucas (int n, int r, int p)
+    {
+        if(r==0)
```
Comment (MastersAbh) [inline Lucas_Theorem/Lucas_Theorem.java]: space around ==

File: Lucas_Theorem/Lucas_Theorem.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+            return 1;
+        int ni,ri;
+
+        ni = n%p;
+        ri = r%p;
+        return (nCrModpLucas(n/p, r/p, p) * nCrModpDP(ni, ri, p)) % p;
+
```
Comment (MastersAbh) [inline Lucas_Theorem/Lucas_Theorem.java]: please add space around operators.

File: Lucas_Theorem/Lucas_Theorem.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+        int n,r,p;
+        Scanner sc = new Scanner(System.in);
+        n = sc.nextInt();
+        r = sc.nextInt();
+        p = sc.nextInt();
+        
+        System.out.println("Value of nCr % p is " + nCrModpLucas(n,r,p));
+    }
+}
+/*      INPUT - OUTPUT
```
Comment (MastersAbh) [inline Lucas_Theorem/Lucas_Theorem.java]: A space before comments.

**Comment by prashant-mahanta** [comment]
@MastersAbh I have fixed the suggested changes have a look at it.

File: Lucas_Theorem/Lucas_Theorem.java [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+public class Lucas_Theorem {
+
+    static int nCrModpDP (int n, int r, int p)
+    {
+        int C[] = new int[r + 1];
+        C[0] = 1;
+
+        for (int i = 1; i <= n; i++)
+        {
+            for(int j = Math.min(i, r); j > 0; j--)
```
Comment (MastersAbh) [inline Lucas_Theorem/Lucas_Theorem.java]: A space after for would be nice

File: Lucas_Theorem/Lucas_Theorem.java (hunk 5 above)
Comment (MastersAbh) [reply Lucas_Theorem/Lucas_Theorem.java]: space around minus

File: Lucas_Theorem/Lucas_Theorem.py [hunk 10]
```diff
@@ -0,0 +1,36 @@
+def nCrModpDP(n,r,p):
+    C=[0]*(r+1)
```
Comment (MastersAbh) [inline Lucas_Theorem/Lucas_Theorem.py]: space around = , *, +

File: Lucas_Theorem/Lucas_Theorem.py [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+    C[0] = 1
+    for i in range(1,n+1):
```
Comment (MastersAbh) [inline Lucas_Theorem/Lucas_Theorem.py]: space around comma, +

File: Lucas_Theorem/Lucas_Theorem.py [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+        for j in range(min(i,r),0,-1):
```
Comment (MastersAbh) [inline Lucas_Theorem/Lucas_Theorem.py]: space after ,

File: Lucas_Theorem/Lucas_Theorem.py [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+            C[j] = (C[j] + C[j-1])%p
```
Comment (MastersAbh) [inline Lucas_Theorem/Lucas_Theorem.py]: space around -, %

File: Lucas_Theorem/Lucas_Theorem.py [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+
+def nCrModpLucas(n,r,p):
+    n, r = int(n), int(r)
+    if r == 0:
+        return 1
+
+    ni = n%p
+    ri = r%p
+    return (nCrModpLucas(n/p, r/p, p) * nCrModpDP(ni, ri, p))%p
+
```
Comment (MastersAbh) [inline Lucas_Theorem/Lucas_Theorem.py]: space around operators.

File: Lucas_Theorem/Lucas_Theorem.py [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+def main():
+    n = int(input())
+    r = int(input())
+    p = int(input())
+    print("Value of nCr % p is",nCrModpLucas(n,r,p))
```
Comment (MastersAbh) [inline Lucas_Theorem/Lucas_Theorem.py]: space after comma

**Comment by prashant-mahanta** [comment]
@MastersAbh Sorry for so many mistakes. Have updated the code. Please have a look at it.

File: Lucas_Theorem/Lucas_Theorem.py [hunk 16]
```diff
@@ -0,0 +1,36 @@
+def nCrModpDP(n, r, p):
+    C=[0] * (r + 1)
```
Comment (MastersAbh) [inline Lucas_Theorem/Lucas_Theorem.py]: some mistake in this line ? 
C=[0] * (r + 1)

**Comment by prashant-mahanta** [comment]
@MastersAbh I have fixed the mistakes.

**Comment by prashant-mahanta** [comment]
@GOVINDDIXIT  Hey, can you please review it. As I have fixed the requested changes. Please  let me know if anything to correct.

**Pull request merged**